Add `call engine julia` commands

jkrumbiegel opened the pull request:
## Description

This PR adds CLI commands to interact with or diagnose the native julia engine. It adds a new subcommand `call engine julia` for this purpose:

```
$ quarto call engine            

Usage:   quarto call engine
Version: 99.9.9       

Description:

  Access functionality specific to quarto's different rendering engines.

Options:

  -h, --help  - Show this help.  

Commands:

  help   [command]  - Show this help or the help of a sub-command.                
  julia             - Access functionality specific to the julia rendering engine.
```

## Implementation

The PR adds a new optional field `populateCommand` to the `ExecutionEngine` interface which only the julia engine currently sets, but others could potentially as well. This field takes a function that adds its own cliffy subcommands under its key, in the julia engine's case `julia`.

## Julia commands

The subcommands for a given engine can be inspected as well. In this PR I've added `status`, `kill` and `log` and `close [--force]`.

```
$ quarto call engine julia

Usage:   quarto call engine julia
Version: 99.9.9             

Description:

  Access functionality specific to the julia rendering engine.

Options:

  -h, --help              - Show this help.                             
  --log         <file>    - Path to log file                            
  --log-level   <level>   - Log level (info, warning, error, critical)  
  --log-format  <format>  - Log format (plain, json-stream)             
  --quiet                 - Suppress console output.                    
  --profile               - Active project profile(s)                   

Commands:

  help    [command]  - Show this help or the help of a sub-command.                               
  status             - Get status information on the currently running Julia server process       
  kill               - Kills the control server if it is currently running.                       
  log                - Prints the julia server log file if it exists which can be used to diagnose
                       problems. 
```

The `status` command can be used to print out information about the currently running server process. For example, if `quarto render scratch/test.qmd` has just been started the output is:

```julia
quarto call engine julia status
Julia server is responding.
  port: 8003
  pid: 48443
  julia version: 1.10.5
  environment: /Users/krumbiegel/.julia/dev/QuartoNotebookRunner
  runner version: 0.11.7
  workers active: 1
    worker 1:
      path: /Users/krumbiegel/dev/quarto-cli/scratch/test.qmd      
      run started: 3:48:01 PM (6 seconds ago)    
      run finished: -    
      timeout: 1000 seconds 
```

or when some notebooks are idling and waiting to be rerun:

```julia
$ quarto call engine julia status
Julia server is responding.
  port: 8003
  pid: 48443
  julia version: 1.10.5
  environment: /Users/krumbiegel/.julia/dev/QuartoNotebookRunner
  runner version: 0.11.7
  workers active: 1
    worker 1:
      path: /Users/krumbiegel/dev/quarto-cli/scratch/test.qmd      
      run started: 3:48:01 PM (1 minute 2 seconds ago)    
      run finished: 3:48:11 PM (took 10 seconds)    
      timeout: 1000 seconds (15 minutes 48 seconds left)
```

Before, there was no way to check what had gone wrong if the server didn't respond anymore. Now, the julia process writes to a log file which can be inspected with the `log` command:

```
$ quarto call engine julia log
[ Info: Log started at 2025-01-06T15:44:19.491
ERROR: LoadError: something failed
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] top-level scope
   @ ~/dev/quarto-cli/src/resources/julia/quartonotebookrunner.jl:92
in expression starting at /Users/krumbiegel/dev/quarto-cli/src/resources/julia/quartonotebookrunner.jl:92
```

If nothing helps or if the server should be restarted for some other reason, the `kill` command sends `SIGTERM` to the `pid` that's saved in the transport file.

```
$ quarto call engine julia kill
Sent SIGTERM to server process
```

The `close` command can be used to close workers that are currently idle. To close workers that are currently computing something, the `--force` flag can be added which adds a way to recover from hanging or long-running worker processes.

## Trying this PR

This PR depends on unreleased changes to QuartoNotebookRunner.jl. You can clone the repository and check out [PR 229](https://github.com/PumasAI/QuartoNotebookRunner.jl/pull/229). Then you can run quarto with `QUARTO_JULIA_PROJECT=path/to/cloned/repo`.

## Summary

This PR adds julia-engine specific commands to the quarto CLI interface. Other than creating the means for other engines to hook in their own commands, this PR does not add any functionality for non-julia engines.

~At the current stage I would first like to get maintainer approval for the general mechanism I introduced here. Comments about the interface for the actual commands is appreciated as well. I will probably also need a little help once the PR goes into the test writing stage, as I have only interacted with the notebook rendering test infrastructure so far, not the general tests for the CLI.~

jkrumbiegel committed:
initial stubs

jkrumbiegel committed:
use dynamically added cliffy subcommands instead

jkrumbiegel committed:
factor out juliaCmd()

jkrumbiegel committed:
remove unused supervisor_pid

jkrumbiegel committed:
refactor json printing and fix race condition in file reading

jkrumbiegel committed:
add kill command and print julia version

jkrumbiegel committed:
print env and runner version

jkrumbiegel committed:
check readiness of server connection for status

jkrumbiegel committed:
use info instead of console.log

jkrumbiegel committed:
reorder log statement

jkrumbiegel committed:
fix transportFile for case where QNR is used from project

jkrumbiegel committed:
refactor server commands

jkrumbiegel committed:
add workers command and per-worker status info

jkrumbiegel committed:
redirect julia server output to logfile

jkrumbiegel committed:
add log to file logic to server process and log command

jkrumbiegel committed:
print seconds since started worker

jkrumbiegel committed:
don't print BEGIN END markers for `log` command itself

jkrumbiegel committed:
Merge branch 'main' into jk/engine-commands

jkrumbiegel committed:
add logfile arg to windows branch

cscheid commented:
I want to give a little bit of further thought to the `engine` name for two reasons:

- a top-level command will live as long as `quarto` itself
- our public top-level commands are all supposed to be verbs

`quarto configure engine` is a bit a of handful to type, but I think it reads much better.

jkrumbiegel commented:
> a top-level command will live as long as quarto itself

I completely understand, which is why I wanted to get your feedback early :)

> our public top-level commands are all supposed to be verbs

I see, that didn't occur to me before. Not sure about `configure engine` given `status` and `log` but might make sense in the bigger picture. Maybe it could be allowed to use `configure julia` directly as a shortcut, then only the internal engine names would have to be avoided if other configurable things are added over time. `configure engine xyz` could be a fallback for future extensions.

jkrumbiegel commented:
@cscheid do you want me to go ahead and switch this PR to `configure engine julia`?

jkrumbiegel committed:
Merge branch 'main' into jk/engine-commands

jkrumbiegel commented:
One more bump :)

jkrumbiegel commented:
About this one, @cscheid, let me recap what my current thinking is here:

- the idea to make engines configurable via CLI was accepted in general
- the exact API to do it was not confirmed yet
  - `quarto engine` seemed to be rejected because `engine` is not a verb
  - maybe `quarto configure engine julia` or `quarto configure julia` would work -> need confirmation on that
- I need help how to add tests for this. So far I've only written smoke tests that check rendering. This would need to test other CLI commands.
- The counterpart of this in QuartoNotebookRunner.jl is not yet merged. I might be able to make CI run off a QNR branch so I can demonstrate things are working before we commit to anything.

cscheid commented:
(Thanks for bumping) `quarto configure engine julia` sounds like it would _change_ something about the engine, but this is about checking, or inspecting.

Proposal: I actually think `quarto check engine julia` is precisely what we would want. It's a bit verbose, but each additional word makes sense given the previous ones, `quarto check` -> `quarto check engine` -> `quarto check engine julia`.

Context:

- Currently, `quarto check` is really about internal health self-checks. But we even have `quarto check jupyter` which starts getting close to what you're looking to do with `quarto check engine julia`.
- In addition, we have `quarto check all` (which is the default for `quarto check`). That could translate easily into `quarto check engine all` being equivalent to `quarto check engine`
- `quarto check` is implemented as a simple dispatch over words. Adding `engine` would be trivial.

So what about the following? We add a new `quarto check engine` subcommand, and then when someone writes `quarto check engine julia`, we dispatch this to the engines through a new method `check(what: string[])`.

The main engineering complication is that this potentially propagates the Cliffy dependency into `check`, and into the engine checks themselves. It would happen because in the same way that someone does `quarto check --help`, they could then want `quarto check engine --help`, `quarto check engine julia --help`, etc.

jkrumbiegel commented:
> quarto configure engine julia sounds like it would change something about the engine, but this is about checking, or inspecting.

The `status` and `log` commands fit this description, but the `kill` one doesn't. And I would potentially add other ones that for example interact with the internal installation of `QuartoNotebookRunner` (`update`) or with the Julia project that quarto would consider the one belonging to a given notebook `quarto ... engine julia resolve notebook.qmd` or something. So these are definitely more changing than checking. For example, currently we are confined to a single QuartoNotebookRunner version, but that doesn't need to be the case if we have an update command.

cscheid commented:
I see. But in that case, `quarto configure engine julia kill` is also weird.

I think I'm getting ready to accept that `quarto engine julia kill` is the least bad of our options.

jkrumbiegel commented:
Yeah I could also not come up with a better verb, yet :) Maybe it's not so bad that engine is not a verb

cscheid commented:
@cderv, @gordonwoodhull, @icarusz Let's come to a decision about this in today's meeting. It's urgent for us to merge this PR, even though it's important to get the name right. I'm currently of the mind that `quarto engine` is the best we have, but I want to make sure everyone is aware of it.

cscheid commented:
@MichaelHatherly the problem is that every other top-level command in quarto is a verb. If you mix it, then it becomes "intuitive" for people to try incorrect commands like (for example) "quarto project create" instead of "quarto create project".

cscheid commented:
@cderv found a _perfect_ verb: `quarto call`. So this would be `quarto call engine julia` ...

@jkrumbiegel The idea would be for your PR to switch from `engine` to `call`, and then the `call` command currently would only support `engine` as a subcommand (although we already have 3 different uses for `quarto call` :)), which would forward the subcommands to the different engines.

Is my description enough to get you going or would you like me to start with an empty `call` command feature, which you would then use to make `call engine`?

jkrumbiegel commented:
Thanks @cscheid, I can switch my PR to that pattern, will push a new version tomorrow, probably.

jkrumbiegel committed:
Merge branch 'main' into jk/engine-commands

jkrumbiegel committed:
switch to `quarto call engine julia ...` pattern

jkrumbiegel committed:
simplify status command by letting QNR do all the work

jkrumbiegel committed:
add first basic call julia engine tests

jkrumbiegel committed:
try different dir for qnr

jkrumbiegel committed:
has to be under workspace it seems

jkrumbiegel committed:
use update instead of instantiate

jkrumbiegel committed:
fix env interpolation

jkrumbiegel committed:
debug windows issue

jkrumbiegel committed:
use bash to set env var on windows also

jkrumbiegel committed:
one more bash shell

jkrumbiegel committed:
assert success with debug info in failure case

jkrumbiegel committed:
add second logfile test and factor out stderr/stdout asserts

jkrumbiegel committed:
type `writeJuliaCommand` better

jkrumbiegel committed:
add close command to close a worker

jkrumbiegel committed:
add test for closing running worker

jkrumbiegel committed:
add command to stop (not kill) the server

jkrumbiegel committed:
period

jkrumbiegel committed:
add log message when server shuts down

jkrumbiegel committed:
use `--force` option for `close` instead of separate command

jkrumbiegel committed:
add test for force-closing a worker

jkrumbiegel committed:
Merge branch 'main' into jk/engine-commands

jkrumbiegel committed:
Merge branch 'main' into jk/engine-commands

jkrumbiegel committed:
remove stale comments

jkrumbiegel committed:
throw only one error and don't triple-print ERROR

jkrumbiegel committed:
remove status info code that is now printed by server on julia side

jkrumbiegel committed:
move julia cache step forward

jkrumbiegel commented:
I'm marking this "ready for review" now because I'm generally happy with the functionality. However, the CI script still pulls in this PR https://github.com/PumasAI/QuartoNotebookRunner.jl/pull/229 which probably only makes sense to merge after this PR here was ok'ed, as they belong together.

I have changed the interface to `quarto call engine julia ...` as requested.

No docs have been written, yet, I will do so once this PR passes review.

cscheid commented:
This is great, and we'll get a lot of mileage out of `quarto call`. Thanks for putting it together!

Could you add an entry to the changelog? Just use the number of this pull request as reference, and use the "Other Fixes and Improvements" section (you'll see examples)

jkrumbiegel committed:
add changelog entries

jkrumbiegel commented:
All right, I have added two entries, one more focused on the julia commands and one on the new `quarto call`

cscheid commented on .github/workflows/test-smokes.yml:
```diff
@@ -174,9 +174,24 @@ jobs:
+      - name: Set QUARTO_JULIA_PROJECT environment variable
+        run: echo "QUARTO_JULIA_PROJECT=$GITHUB_WORKSPACE/qnr" >> $GITHUB_ENV
+        shell: bash
+
```
Do I understand correctly that this would eventually be changed to check out the `main` brtanch of QuartoNotebookRunner once that's released?

The question I have is: will our users have to check out this branch in order for the feature to work?

jkrumbiegel replied:
the PR with those changes is waiting for review over there, given that you are positive about the changes made here, we will probably go ahead with that in a little bit (cc @MichaelHatherly). Then, we can simply change the version number in the Project.toml file like we usually do and replace this CI script modification.

MichaelHatherly replied:
Yeah, I'll be looking at the backend PR this afternoon.

MichaelHatherly replied:
Backend PR looks fine, @jkrumbiegel feel free to merge and tag a new version whenever you're ready.

cscheid commented on src/command/call/cmd.ts:
```diff
@@ -0,0 +1,9 @@
+  .action(() => {
+    callCommand.showHelp();
+    Deno.exit(1);
+  }).command("engine", engineCommand);
```
If I understand this correctly, we will need a `description(...)` here. Otherwise `quarto call --help` doesn't do anything, right?

It'll necessarily be something generic like "access functions of Quarto subsystems such as 'engine'", and we'll update it as appropriate.

cscheid commented on src/execute/engine.ts:
```diff
@@ -276,3 +277,30 @@ export function projectIgnoreGlobs(dir: string) {
+    engineSubcommand
+      .description(
+        `Access functionality specific to the ${name} rendering engine.`,
+      )
```
I like this. I think that we should let engines have a method to build a response, but that can happen in a different PR.

cscheid reviewed (approved):
I'm going to approve this. With that said, unless I'm missing something, we should consider adding a `description()` to `src/command/call/cmd.ts`

jkrumbiegel committed:
switch to qnr 0.15 and remove custom branch from CI workflow

jkrumbiegel committed:
add description for `call`

jkrumbiegel committed:
slightly reword julia command help strings

jkrumbiegel commented:
All right, I've released QNR 0.15 and changed compat to that version so that I could remove the special branch from CI again. I've also added a description to `call` and made some further small changes to help strings. With that, I think this PR is good to merge if CI goes green once more.

jkrumbiegel committed:
Merge branch 'main' into jk/engine-commands

cscheid commented:
🎉

cderv commented:
Thank you @jkrumbiegel !

jkrumbiegel commented:
Thank you for merging, I'll write up some docs next week

jkrumbiegel commented:
I have added docs here https://github.com/quarto-dev/quarto-web/pull/1591
